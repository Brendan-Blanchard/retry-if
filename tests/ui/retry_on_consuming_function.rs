use retry_if::{retry, ExponentialBackoffConfig};
use std::time::Duration;

const BACKOFF_CONFIG: ExponentialBackoffConfig = ExponentialBackoffConfig {
    max_retries: 5,
    t_wait: Duration::from_secs(1),
    backoff: 2.0,
    t_wait_max: None,
    backoff_max: None,
};

fn retry_if(_i: &NewCounter) -> bool {
    true
}

pub struct Counter {
    pub count: i64,
}

pub struct NewCounter {
    pub count: i64,
}

impl From<Counter> for NewCounter {
    fn from(value: Counter) -> Self {
        NewCounter {
            count: value.count,
        }
    }
}

impl Counter {
    #[retry(BACKOFF_CONFIG, retry_if)]
    async fn consume_self(self) -> NewCounter {
        self.into()
    }
}

#[tokio::main]
async fn main() {
    let counter = Counter { count: 0 };
    let _new_counter = counter.consume_self().await;
}