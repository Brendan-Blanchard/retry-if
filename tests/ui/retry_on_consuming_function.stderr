error[E0382]: use of moved value: `self`
  --> tests/ui/retry_on_consuming_function.rs:35:9
   |
33 |     #[retry(BACKOFF_CONFIG, retry_if)]
   |     ---------------------------------- inside of this loop
34 |     async fn consume_self(self) -> NewCounter {
   |                           ---- move occurs because `self` has type `Counter`, which does not implement the `Copy` trait
35 |         self.into()
   |         ^^^^ ------ `self` moved due to this method call, in previous iteration of loop
   |
note: `into` takes ownership of the receiver `self`, which moves `self`
  --> $RUST/core/src/convert/mod.rs
